### PR TITLE
[FIP-38] Tokemak deposit

### PR DESCRIPTION
**Pull request by eswak**
This pull request implements smart contracts for depositing ETH & ERC20s on Tokemak.

- [x] Implement smart contracts
- [x] Unit tests & Mock contracts
- [x] Integration tests on a local mainnet fork
- [x] DAO description
- [x] DAO script

image

### CODE REVIEW

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol
Comment (Joeysantoro) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:42]: can be immutable

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol
Comment (Joeysantoro) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:48]: can be immutable

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol
Comment (Joeysantoro) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:45]: probably can be immutable?

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol
Comment (Joeysantoro) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:86]: Should emit an event

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol
Comment (Joeysantoro) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:102]: chainId can probably be hardcoded, is the wallet this address? That can also be hardcoded if yes

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol
Comment (Joeysantoro) [inline contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:36]: Does this need to be approved?

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol
Comment (Joeysantoro) [inline contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:46]: Is there an edge case around not having requested a withdrawal?

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts
Comment (Joeysantoro) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:44]: We've been using state variables with local scope to take advantage of typescript types recently on the tests

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts
Comment (Joeysantoro) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:1]: Can use `@test/helpers`. Typescript aliases paths for us

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts
Comment (Joeysantoro) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:53]: Here and in many places there are empty braces as a param, these can be removed I think

File: test/unit/pcv/EthTokemakPCVDeposit.test.ts
Comment (Joeysantoro) [inline test/unit/pcv/EthTokemakPCVDeposit.test.ts:8]: same comments as ERC20 one

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts
Comment (ditchfieldcaleb) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:123]: (see other comments about expect revert syntax)

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts
Comment (ditchfieldcaleb) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:92]: (see other comments about expect revert syntax)

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts
Comment (ditchfieldcaleb) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:74]: (see other comments about expect revert syntax)

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts
Comment (ditchfieldcaleb) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:54]: (see other comments about expect revert syntax)

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts
Comment (ditchfieldcaleb) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:39]: (see other comment about `this` in test files)

**Comment by Joeysantoro** [comment]
Overall looks really great, you can fix these nits then send to Tokemak team for review. Also feel free to begin DAO scripts

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol
Comment (ditchfieldcaleb) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:121]: should declare this as a constant or immutable var instead of hardcoded here

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol
Comment (ditchfieldcaleb) [inline contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:36]: What's this approval do/for?

**Review by ditchfieldcaleb — CHANGES_REQUESTED** [review]
Contracts look great overall, just a couple of questions there. Roles all look fine and love the comments.

For the tests I left a lot of comments; most of them are code-style things (we've updated the way we do tests quite a bit in the past few months).

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol
Comment (eswak) [reply contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:46]: The Tokemak ETH pool will revert if we have not called `requestWithdrawal` before : https://github.com/Tokemak/tokemak-smart-contracts-public/blob/main/contracts/pools/EthPool.sol#L94-L97

Same for the Tokemak ERC20 pool : https://github.com/Tokemak/tokemak-smart-contracts-public/blob/main/contracts/pools/Pool.sol#L87-L90

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol
Comment (eswak) [reply contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:36]: not for the ETH deposit, good catch :)

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol
Comment (eswak) [reply contracts/pcv/tokemak/TokemakPCVDepositBase.sol:102]: yes, I'll hardcode both

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts
Comment (eswak) [reply test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:169]: in the unit test that use the mock rewards contract, it will always work. I removed the address parameter (joey review), so I'll mark this comment as resolved

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol
Comment (eswak) [reply contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:36]: removed, WETH does not need approval (pure ETH is passed to the payable function), good catch :)

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol
Comment (eswak) [reply contracts/pcv/tokemak/TokemakPCVDepositBase.sol:121]: Added as : `address private constant TOKE_TOKEN_ADDRESS = address(0x2e9d63788249371f1DFC918a52f8d799F4a38C94);`

**Review by Joeysantoro — APPROVED** [review]

**Review by ditchfieldcaleb — DISMISSED** [review]

File: proposals/dao/fip_38.ts
Comment (Joeysantoro) [inline proposals/dao/fip_38.ts:19]: These should probably be added to mainnetAddresses.json right?

File: scripts/deploy/fip_38.ts
Comment (Joeysantoro) [inline scripts/deploy/fip_38.ts:6]: This file is not needed

**Comment by Joeysantoro** [comment]
After fixing merge conflicts and pulling in develop should be good to 🚢

File: proposals/dao/fip_38.ts
Comment (eswak) [reply proposals/dao/fip_38.ts:19]: I just needed them here, but I can add them. What `artifactName` do I use ?

**Comment by eswak** [comment]
fixed the e2e tests, didn't know about the `dependencies.ts` file 😅

**Pull request merged**
